src/main/java/org/truffleruby/core/regexp/MatchDataGuards.java

```diff
@@ -0,0 +1,20 @@
+/*
+ * Copyright (c) 2015, 2019 Oracle and/or its affiliates. All rights reserved. This
+ * code is released under a tri EPL/GPL/LGPL license. You can use it,
+ * redistribute it and/or modify it under the terms of the:
+ *
+ * Eclipse Public License version 2.0, or
+ * GNU General Public License version 2, or
+ * GNU Lesser General Public License version 2.1.
+ */
+
+package org.truffleruby.core.regexp;
+
+import com.oracle.truffle.api.object.DynamicObject;
+import org.truffleruby.Layouts;
+
+public class MatchDataGuards {
+    public static boolean isInitialized(DynamicObject matchData) {
+        return Layouts.MATCH_DATA.getRegexp(matchData) != null;
```

Comment:
Could you check `getSource(matchData) != null` here?
`getRegexp()` can already return multiple kinds of values (see comment).